File: src/infrastructure/provider/anthropic.ts
/**
 * Anthropic Provider — 使用 Anthropic Messages API 格式
 *
 * 与 OpenAI（content 为字符串/tool_calls 在顶层）不同：
 * - content 始终为 block array，tool_use/tool_result 嵌入 content 内
 * - 流式输出包含 content_block_delta / content_block_stop 事件
 * - 使用 x-api-key header
 * - API 端点：https://api.anthropic.com/v1/messages
 */
import type {
  Message,
  ProviderConfig,
  StreamCallback,
  ToolUseContent,
} from '@/types'
import type { ChatRequest, IProvider } from './types'
import { apiFetch, getResponseReader, readStreamLines } from './http-utils'
import { getLastSummaryMessageIndex } from '@/types'
import { processVisionContent } from './visionInject'
import { fetch } from '@tauri-apps/plugin-http'

interface AnthropicContentBlock {
  type: string
  text?: string
  id?: string
  name?: string
  input?: Record<string, any>
  tool_use_id?: string
  content?: string
  source?: { type: string; media_type?: string; data?: string; url?: string }
  is_error?: boolean
  thinking?: string
  signature?: string
}

interface AnthropicResponse {
  id: string
  type: string
  role: string
  content: AnthropicContentBlock[]
  model: string
  stop_reason: string | null
  usage?: {
    input_tokens: number
    output_tokens: number
    cache_read_input_tokens?: number
    cache_creation_input_tokens?: number
  }
}
export class AnthropicProvider implements IProvider {
  readonly name: string
  private apiKey: string
  private baseUrl: string

  private get headers() {
    return {
      'Content-Type': 'application/json',
      'x-api-key': this.apiKey,
      'anthropic-version': '2023-06-01',
    }
  }

  constructor(name: string, apiKey: string, baseUrl?: string) {
    this.name = name
    this.apiKey = apiKey
    this.baseUrl = (baseUrl || 'https://api.anthropic.com/v1').replace(
      /\/+$/,
      '',
    )
  }
  async validateApiKey(config: ProviderConfig): Promise<boolean> {
    if (config.models.length === 0) {
      throw new Error('请先为该 Provider 配置至少一个模型')
    }
    try {
      const res = await fetch(`${this.baseUrl}/messages`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'x-api-key': this.apiKey,
          'anthropic-version': '2023-06-01',
        },
        body: JSON.stringify({
          model: config.models[0],
          messages: [
            { role: 'user', content: [{ type: 'text', text: 'ping' }] },
          ],
          max_tokens: 1,
        }),
      })
      if (!res.ok) {
        throw new Error(await res.text())
      }
      return res.ok
    } catch (e) {
      console.error(e)
      return false
    }
  }

  async listModels(): Promise<string[]> {
    try {
      const urlObj = new URL(this.baseUrl)
      const res = await fetch(`${urlObj.protocol}//${urlObj.host}/models`, {
        headers: {
          Authorization: `Bearer ${this.apiKey}`,
          'x-api-key': this.apiKey,
          'anthropic-version': '2023-06-01',
        },
      })
      if (!res.ok) throw '请手动填写模型'
      const data = await res.json()
      return (data.data || []).map((m: any) => m.id)
    } catch {
      throw '请手动填写模型'
    }
  }

  async chat(request: ChatRequest, signal?: AbortSignal): Promise<Message> {
    const body = this.buildRequest(request)
    const res = await apiFetch({
      url: `${this.baseUrl}/messages`,
      headers: this.headers,
      body: JSON.stringify(body),
      signal,
      providerName: 'Anthropic',
    })
    const data: AnthropicResponse = await res.json()
    return this.parseResponse(data)
  }

  async chatStream(
    request: ChatRequest,
    callback: StreamCallback,
    signal?: AbortSignal,
  ): Promise<void> {
    const body = this.buildRequest(request)
    body.stream = true

    let res: Response
    try {
      res = await apiFetch({
        url: `${this.baseUrl}/messages`,
        headers: this.headers,
        body: JSON.stringify(body),
        signal,
        providerName: 'Anthropic',
      })
    } catch (e: any) {
      callback({ type: 'error', error: e.message })
      return
    }

    let reader: ReadableStreamDefaultReader<Uint8Array>
    try {
      reader = getResponseReader(res)
    } catch (e: any) {
      callback({ type: 'error', error: e.message })
      return
    }

    const decoder = new TextDecoder()
    await this.parseAnthropicSSE(reader, decoder, callback, signal)
  }

  private async parseAnthropicSSE(
    reader: ReadableStreamDefaultReader<Uint8Array>,
    decoder: TextDecoder,
    callback: StreamCallback,
    signal: AbortSignal | undefined,
  ): Promise<void> {
    let currentEvent = ''
    const blockTexts: Map<number, string> = new Map()
    const toolUses: Map<number, ToolUseContent> = new Map()
    const inputPartials: Map<number, string> = new Map()
    let toolUseEventFired = false
    let thinkingContentBuffer = ''
    let lastUsage:
      | { promptTokens: number; completionTokens: number; totalTokens: number }
      | undefined

    try {
      await readStreamLines(
        reader,
        decoder,
        (line) => {
          const trimmed = line.trim()

          if (trimmed.startsWith('event:')) {
            currentEvent = trimmed.slice(6).trim()
            return
          }

          if (trimmed.startsWith('data:')) {
            const dataStr = trimmed.slice(5).trim()

            if (dataStr === '[DONE]') {
              callback({ type: 'message_stop' })
              return
            }

            try {
              const data = JSON.parse(dataStr)

              switch (currentEvent) {
                case 'content_block_start': {
                  if (data.content_block?.type === 'tool_use') {
                    toolUses.set(data.index, {
                      type: 'tool_use',
                      id: data.content_block.id,
                      name: data.content_block.name,
                      input: data.content_block.input || {},
                    })
                  }
                  blockTexts.set(data.index, '')
                  break
                }

                case 'content_block_delta': {
                  const delta = data.delta
                  if (delta?.type === 'text_delta') {
                    const current = blockTexts.get(data.index) || ''
                    blockTexts.set(data.index, current + delta.text)
                    callback({ type: 'text_delta', data: delta.text })
                  }
                  if (delta?.type === 'input_json_delta') {
                    const existing = inputPartials.get(data.index) || ''
                    inputPartials.set(
                      data.index,
                      existing + (delta.partial_json || ''),
                    )
                  }
                  if (delta?.type === 'thinking_delta') {
                    thinkingContentBuffer += delta.thinking || ''
                    callback({
                      type: 'reasoning_content_change',
                      data: thinkingContentBuffer,
                    })
                  }
                  break
                }

                case 'content_block_stop': {
                  const toolUse = toolUses.get(data.index)
                  if (toolUse) {
                    const partial = inputPartials.get(data.index)
                    if (partial) {
                      try {
                        toolUse.input = JSON.parse(partial)
                      } catch {
                        toolUse.input = { _partial: partial }
                      }
                    }
                    if (!toolUseEventFired) {
                      toolUseEventFired = true
                      callback({ type: 'tool_use', toolUse })
                    }
                  }
                  break
                }

                case 'message_delta': {
                  if (data.usage) {
                    lastUsage = {
                      promptTokens: data.usage.input_tokens ?? 0,
                      completionTokens: data.usage.output_tokens ?? 0,
                      totalTokens:
                        (data.usage.input_tokens ?? 0) +
                        (data.usage.output_tokens ?? 0) +
                        (data.usage.cache_read_input_tokens ?? 0) +
                        (data.usage.cache_creation_input_tokens ?? 0),
                    }
                  }
                  if (
                    data.delta?.stop_reason === 'tool_use' &&
                    !toolUseEventFired
                  ) {
                    for (const [, tu] of toolUses) {
                      callback({ type: 'tool_use', toolUse: tu })
                    }
                    toolUseEventFired = true
                  }
                  break
                }

                case 'message_stop': {
                  callback({
                    type: 'message_stop',
                    reasoningContent: thinkingContentBuffer || undefined,
                    usage: lastUsage,
                  })
                  break
                }

                case 'error': {
                  callback({
                    type: 'error',
                    error: data.error?.message || 'Anthropic API error',
                  })
                  break
                }
              }
            } catch {
              // Parse error
            }
            return
          }

          // Empty line resets event type
          if (trimmed === '') {
            currentEvent = ''
          }
        },
        signal,
      )
    } catch (e: any) {
      if (e?.name !== 'AbortError') {
        callback({ type: 'error', error: e.message })
      }
    }
  }

  buildRequest(request: ChatRequest): any {
    // Anthropic API: system is a top-level field, messages are [{role, content}]
    // content is always an array of blocks

    const messages: { role: string; content: AnthropicContentBlock[] }[] = []
    let system: string | undefined = request.systemPrompt || ''

    // 收集当前消息批次中已完成内容的索引
    let tempAssistantBlocks: AnthropicContentBlock[] = []
    const lastSummaryMessageIndex = getLastSummaryMessageIndex(request.messages)
    const requestMessages =
      lastSummaryMessageIndex == -1
        ? request.messages
        : request.messages.slice(lastSummaryMessageIndex)
    for (const msg of requestMessages) {
      // summary 角色：转为 user 消息，作为压缩后的历史上下文
      if (msg.role === 'summary') {
        messages.push({
          role: 'user',
          content: [
            {
              type: 'text',
              text:
                typeof msg.content === 'string'
                  ? `${msg.content}`
                  : '' + JSON.stringify(msg.content),
            },
          ],
        })
        continue
      }

      if (msg.role === 'assistant') {
        tempAssistantBlocks = []
        const content = msg.content
        // 将历史消息中的 thinking 内容作为 thinking block 传给 API
        if (msg.reasoningContent) {
          tempAssistantBlocks.push({
            type: 'thinking',
            thinking: msg.reasoningContent,
          })
        }
        if (typeof content === 'string' && content) {
          tempAssistantBlocks.push({ type: 'text', text: content })
        }

        if (msg.toolCalls?.length) {
          for (const tc of msg.toolCalls) {
            tempAssistantBlocks.push({
              type: 'tool_use',
              id: tc.id,
              name: tc.name,
              input: tc.input,
            })
          }
        }

        messages.push({
          role: 'assistant',
          content:
            tempAssistantBlocks.length > 0
              ? tempAssistantBlocks
              : [{ type: 'text', text: '' }],
        })
        continue
      }

      if (msg.role === 'tool') {
        // tool_result — content is a block array
        const toolResultBlock: AnthropicContentBlock = {
          type: 'tool_result',
          tool_use_id: msg.toolCallId || '',
          content:
            typeof msg.content === 'string'
              ? msg.content
              : JSON.stringify(msg.content),
          is_error: msg.isError ? true : undefined,
        }

        const lastMsg = messages[messages.length - 1]
        if (
          lastMsg &&
          lastMsg.role === 'user' &&
          lastMsg.content.length > 0 &&
          lastMsg.content[0].type === 'tool_result'
        ) {
          // 前一条消息已经是 tool_result 的 user 消息 → 追加到同一条
          lastMsg.content.push(toolResultBlock)
        } else {
          // 创建新的 tool_result 消息
          messages.push({
            role: 'user',
            content: [toolResultBlock],
          })
        }
        continue
      }

      // user message
      const contentBlocks: AnthropicContentBlock[] = []
      if (typeof msg.content === 'string') {
        contentBlocks.push({ type: 'text', text: msg.content })
      } else if (Array.isArray(msg.content)) {
        for (const block of msg.content) {
          if (block.type === 'text') {
            contentBlocks.push({ type: 'text', text: block.text })
          } else if (block.type === 'image_url') {
            const url = block.image_url.url
            if (url.startsWith('data:')) {
              const parts = url.split(',')
              contentBlocks.push({
                type: 'image',
                source: {
                  type: 'base64',
                  media_type: 'image/jpeg',
                  data: parts[1],
                },
              })
            } else {
              contentBlocks.push({
                type: 'image',
                source: { type: 'url', url },
              })
            }
          }
        }
      }

      // 注入视觉分析结果
      // 视觉分析优化：替换 image_url 为分析文本
      const visionBlocks = processVisionContent(msg)
      if (visionBlocks) {
        // 整个 content 由 vision 接管（去掉 image_url，追加分析文本）
        messages.push({ role: 'user', content: visionBlocks as any })
      } else {
        messages.push({ role: 'user', content: contentBlocks })
      }
    }

    const body: any = {
      model: request.model,
      max_tokens: request.maxTokens ?? 2000000,
      messages,
    }

    if (system) {
      body.system = system.trim()
    }

    if (request.temperature !== undefined) {
      body.temperature = request.temperature
    }

    if (request.tools?.length) {
      body.tools = request.tools.map((t) => ({
        name: t.name,
        description: t.description,
        input_schema: t.parameters,
      }))
    }
    if (request.tool_choice == 'none') {
      body.tool_choice = { type: 'none' }
    }

    return body
  }

  private parseResponse(data: AnthropicResponse): Message {
    const message: Message = {
      id: data.id,
      role: 'assistant',
      content: '',
      timestamp: Date.now(),
    }

    const textBlocks: string[] = []
    const toolCalls: ToolUseContent[] = []

    for (const block of data.content || []) {
      if (block.type === 'text') {
        textBlocks.push(block.text || '')
      } else if (block.type === 'tool_use') {
        toolCalls.push({
          type: 'tool_use',
          id: block.id || '',
          name: block.name || '',
          input: block.input || {},
        })
      }
    }

    message.content = textBlocks.join('')

    if (toolCalls.length > 0) {
      message.toolCalls = toolCalls
    }

    if (data.usage) {
      message.usage = {
        promptTokens: data.usage.input_tokens,
        completionTokens: data.usage.output_tokens,
        totalTokens:
          data.usage.input_tokens +
          data.usage.output_tokens +
          (data.usage.cache_read_input_tokens ?? 0) +
          (data.usage.cache_creation_input_tokens ?? 0),
      }
    }

    return message
  }
}

File: src-tauri/src/agent/provider/anthropic.rs
//! Anthropic Messages API Provider
//!
//! 移植自 TS `src/infrastructure/provider/anthropic.ts`（铁律 1：双引擎同语义）。

use super::blocks::{anthropic_blocks, process_vision_content, slice_messages, text_of_content};
use super::sse::read_sse_lines;
use super::Provider;
use super::super::cancellation::CancellationToken;
use super::super::types::{ChatRequest, Message, StreamEvent, TokenUsage, ToolUseContent};
use async_trait::async_trait;
use serde_json::{json, Value};

// ==================== Anthropic Provider ====================

pub struct NativeAnthropicProvider {
    api_key: String,
    base_url: String,
    http: reqwest::Client,
}

impl NativeAnthropicProvider {
    pub fn new(_name: &str, api_key: &str, base_url: &str) -> Self {
        let base = if base_url.is_empty() {
            "https://api.anthropic.com/v1"
        } else {
            base_url
        };
        Self {
            api_key: api_key.to_string(),
            base_url: base.trim_end_matches('/').to_string(),
            http: reqwest::Client::new(),
        }
    }

    fn headers(&self) -> reqwest::header::HeaderMap {
        let mut h = reqwest::header::HeaderMap::new();
        h.insert(
            reqwest::header::CONTENT_TYPE,
            reqwest::header::HeaderValue::from_static("application/json"),
        );
        h.insert(
            "x-api-key",
            reqwest::header::HeaderValue::from_str(&self.api_key)
                .unwrap_or_else(|_| reqwest::header::HeaderValue::from_static("")),
        );
        h.insert(
            "anthropic-version",
            reqwest::header::HeaderValue::from_static("2023-06-01"),
        );
        h
    }

    /// 构建 Anthropic 请求体（移植 anthropic.ts buildRequest）
    pub(super) fn build_request(&self, request: &ChatRequest) -> Value {
        let mut messages: Vec<Value> = Vec::new();
        let system = request.system_prompt.clone().unwrap_or_default();

        let request_messages = slice_messages(&request.messages);
        for msg in request_messages {
            if msg.role == "summary" || msg.role == "feedback" {
                messages.push(json!({
                    "role": "user",
                    "content": [{
                        "type": "text",
                        "text": text_of_content(&msg.content),
                    }],
                }));
                continue;
            }

            if msg.role == "assistant" {
                let mut blocks: Vec<Value> = Vec::new();
                if let Some(rc) = &msg.reasoning_content {
                    if !rc.is_empty() {
                        blocks.push(json!({ "type": "thinking", "thinking": rc }));
                    }
                }
                if let Value::String(s) = &msg.content {
                    if !s.is_empty() {
                        blocks.push(json!({ "type": "text", "text": s }));
                    }
                }
                if let Some(tcs) = &msg.tool_calls {
                    for tc in tcs {
                        blocks.push(json!({
                            "type": "tool_use",
                            "id": tc.id,
                            "name": tc.name,
                            "input": tc.input,
                        }));
                    }
                }
                if blocks.is_empty() {
                    blocks.push(json!({ "type": "text", "text": "" }));
                }
                messages.push(json!({ "role": "assistant", "content": blocks }));
                continue;
            }

            if msg.role == "tool" {
                let tool_result_block = json!({
                    "type": "tool_result",
                    "tool_use_id": msg.tool_call_id.clone().unwrap_or_default(),
                    "content": text_of_content(&msg.content),
                    "is_error": msg.is_error.unwrap_or(false),
                });
                // 若上一条是 tool_result user 消息则追加
                if let Some(last) = messages.last_mut() {
                    let is_tool_result_user = last.get("role").and_then(Value::as_str) == Some("user")
                        && last
                            .get("content")
                            .and_then(Value::as_array)
                            .and_then(|a| a.first())
                            .and_then(|b| b.get("type"))
                            .and_then(Value::as_str)
                            == Some("tool_result");
                    if is_tool_result_user {
                        if let Some(arr) = last.get_mut("content").and_then(|v| v.as_array_mut()) {
                            arr.push(tool_result_block);
                        }
                        continue;
                    }
                }
                messages.push(json!({ "role": "user", "content": [tool_result_block] }));
                continue;
            }

            // user message
            let blocks: Vec<Value> = match &msg.content {
                Value::String(s) => vec![json!({ "type": "text", "text": s })],
                Value::Array(arr) => anthropic_blocks(arr),
                _ => Vec::new(),
            };
            // 本地图片伪视觉分析：将图片替换为分析文本（纯文本模型不支持 image 类型）
            // vision 重建后的块同样要过降级，否则 file / quote 块会漏给 Anthropic
            let blocks = match process_vision_content(msg) {
                Some(processed) => anthropic_blocks(
                    processed.as_array().map(Vec::as_slice).unwrap_or(&[]),
                ),
                None => blocks,
            };
            messages.push(json!({ "role": "user", "content": blocks }));
        }

        let mut body = json!({
            "model": request.model,
            "max_tokens": request.max_tokens,
            "messages": messages,
        });
        if !system.trim().is_empty() {
            body["system"] = Value::String(system.trim().to_string());
        }
        body["temperature"] = Value::from(request.temperature);
        if !request.tools.is_empty() {
            let tools: Vec<Value> = request
                .tools
                .iter()
                .map(|t| {
                    json!({
                        "name": t.name,
                        "description": t.description,
                        "input_schema": t.parameters,
                    })
                })
                .collect();
            body["tools"] = Value::Array(tools);
        }
        if request.tool_choice == "none" {
            body["tool_choice"] = json!({ "type": "none" });
        }
        body
    }

    fn parse_response(&self, data: &Value) -> Message {
        let mut message = Message {
            id: data.get("id").and_then(Value::as_str).unwrap_or_default().to_string(),
            role: "assistant".to_string(),
            content: Value::String(String::new()),
            tool_calls: None,
            reasoning_content: None,
            tool_call_id: None,
            is_error: None,
            elapsed_ms: None,
            reasoning_elapsed_ms: None,
            ui_data: None,
            timestamp: chrono::Utc::now().timestamp_millis(),
            streaming: None,
            model: None,
            usage: None,
            image_vision_analyze_optimize: None,
            image_vision_analyze_result: None,
        };

        let mut texts: Vec<String> = Vec::new();
        let mut tool_calls: Vec<ToolUseContent> = Vec::new();

        if let Some(blocks) = data.get("content").and_then(Value::as_array) {
            for block in blocks {
                match block.get("type").and_then(Value::as_str) {
                    Some("text") => texts.push(block.get("text").and_then(Value::as_str).unwrap_or("").to_string()),
                    Some("tool_use") => tool_calls.push(ToolUseContent {
                        type_: "tool_use".into(),
                        id: block.get("id").and_then(Value::as_str).unwrap_or("").to_string(),
                        name: block.get("name").and_then(Value::as_str).unwrap_or("").to_string(),
                        input: block.get("input").cloned().unwrap_or(Value::Null),
                    }),
                    _ => {}
                }
            }
        }

        message.content = Value::String(texts.join(""));

        if !tool_calls.is_empty() {
            message.tool_calls = Some(tool_calls);
        }

        if let Some(usage) = data.get("usage") {
            let input = usage.get("input_tokens").and_then(Value::as_i64).unwrap_or(0);
            let output = usage.get("output_tokens").and_then(Value::as_i64).unwrap_or(0);
            let cache_read = usage.get("cache_read_input_tokens").and_then(Value::as_i64).unwrap_or(0);
            let cache_create = usage.get("cache_creation_input_tokens").and_then(Value::as_i64).unwrap_or(0);
            message.usage = Some(TokenUsage {
                prompt_tokens: input,
                completion_tokens: output,
                total_tokens: input + output + cache_read + cache_create,
                cached_tokens: Some(cache_read + cache_create),
            });
        }

        message
    }
}

#[async_trait]
impl Provider for NativeAnthropicProvider {
    async fn chat(
        &self,
        request: &ChatRequest,
        cancel: &CancellationToken,
    ) -> Result<Message, String> {
        let body = self.build_request(request);
        let url = format!("{}/messages", self.base_url);

        let resp = tokio::select! {
            _ = cancel.cancelled() => return Err("cancelled".into()),
            r = self.http.post(&url).headers(self.headers()).json(&body).send() => r.map_err(|e| format!("API Error: {}", e))?,
        };
        let status = resp.status();
        let text = resp.text().await.map_err(|e| format!("读取响应失败: {}", e))?;
        if !status.is_success() {
            return Err(format!("API Error ({}): {}", status.as_u16(), text));
        }
        let data: Value = serde_json::from_str(&text).map_err(|e| format!("响应解析失败: {}", e))?;
        Ok(self.parse_response(&data))
    }

    async fn chat_stream(
        &self,
        request: &ChatRequest,
        cancel: &CancellationToken,
        on_event: &mut (dyn FnMut(StreamEvent) + Send),
    ) -> Result<(), String> {
        let mut body = self.build_request(request);
        body["stream"] = Value::Bool(true);
        let url = format!("{}/messages", self.base_url);

        let resp = tokio::select! {
            _ = cancel.cancelled() => return Err("cancelled".into()),
            r = self.http.post(&url).headers(self.headers()).json(&body).send() => r.map_err(|e| format!("API Error: {}", e))?,
        };
        let status = resp.status();
        if !status.is_success() {
            let text = resp.text().await.map_err(|e| format!("读取错误响应失败: {}", e))?;
            return Err(format!("API Error ({}): {}", status.as_u16(), text));
        }

        let mut current_event = String::new();
        let mut block_texts: std::collections::HashMap<usize, String> = Default::default();
        let mut tool_uses: std::collections::HashMap<usize, ToolUseContent> = Default::default();
        let mut input_partials: std::collections::HashMap<usize, String> = Default::default();
        let mut tool_fired = false;
        let mut thinking_buffer = String::new();
        let mut last_usage: Option<TokenUsage> = None;

        let result = read_sse_lines(resp, cancel, &mut |line: String| {
            let trimmed = line.trim();

            if trimmed.starts_with("event:") {
                current_event = trimmed[6..].trim().to_string();
                return true;
            }

            if trimmed.starts_with("data:") {
                let data_str = trimmed[5..].trim().to_string();
                if data_str == "[DONE]" {
                    on_event(StreamEvent::MessageStop {
                        reasoning_content: if thinking_buffer.is_empty() {
                            None
                        } else {
                            Some(thinking_buffer.clone())
                        },
                        usage: last_usage.clone(),
                    });
                    return true;
                }
                let data: Value = match serde_json::from_str(&data_str) {
                    Ok(v) => v,
                    Err(_) => return true,
                };
                let index = data.get("index").and_then(Value::as_u64).unwrap_or(0) as usize;

                match current_event.as_str() {
                    "content_block_start" => {
                        if let Some(cb) = data.get("content_block") {
                            if cb.get("type").and_then(Value::as_str) == Some("tool_use") {
                                tool_uses.insert(
                                    index,
                                    ToolUseContent {
                                        type_: "tool_use".into(),
                                        id: cb.get("id").and_then(Value::as_str).unwrap_or("").to_string(),
                                        name: cb.get("name").and_then(Value::as_str).unwrap_or("").to_string(),
                                        input: cb.get("input").cloned().unwrap_or(json!({})),
                                    },
                                );
                            }
                        }
                        block_texts.insert(index, String::new());
                    }
                    "content_block_delta" => {
                        if let Some(delta) = data.get("delta") {
                            match delta.get("type").and_then(Value::as_str) {
                                Some("text_delta") => {
                                    let text = delta.get("text").and_then(Value::as_str).unwrap_or("");
                                    block_texts.entry(index).or_default().push_str(text);
                                    on_event(StreamEvent::TextDelta(text.to_string()));
                                }
                                Some("input_json_delta") => {
                                    let partial = delta.get("partial_json").and_then(Value::as_str).unwrap_or("");
                                    input_partials.entry(index).or_default().push_str(partial);
                                }
                                Some("thinking_delta") => {
                                    let thinking = delta.get("thinking").and_then(Value::as_str).unwrap_or("");
                                    thinking_buffer.push_str(thinking);
                                    on_event(StreamEvent::ReasoningContentChange(thinking_buffer.clone()));
                                }
                                _ => {}
                            }
                        }
                    }
                    "content_block_stop" => {
                        if let Some(tool_use) = tool_uses.get_mut(&index) {
                            if let Some(partial) = input_partials.get(&index) {
                                tool_use.input = serde_json::from_str(partial).unwrap_or_else(|_| {
                                    json!({ "_partial": partial })
                                });
                            }
                            if !tool_fired {
                                tool_fired = true;
                                on_event(StreamEvent::ToolUse(tool_use.clone()));
                            }
                        }
                    }
                    "message_delta" => {
                        if let Some(usage) = data.get("usage") {
                            let input = usage.get("input_tokens").and_then(Value::as_i64).unwrap_or(0);
                            let output = usage.get("output_tokens").and_then(Value::as_i64).unwrap_or(0);
                            let cache_read = usage.get("cache_read_input_tokens").and_then(Value::as_i64).unwrap_or(0);
                            let cache_create = usage.get("cache_creation_input_tokens").and_then(Value::as_i64).unwrap_or(0);
                            last_usage = Some(TokenUsage {
                                prompt_tokens: input,
                                completion_tokens: output,
                                total_tokens: input + output + cache_read + cache_create,
                                cached_tokens: Some(cache_read + cache_create),
                            });
                        }
                        if !tool_fired {
                            let stop_reason = data
                                .get("delta")
                                .and_then(|d| d.get("stop_reason"))
                                .and_then(Value::as_str);
                            if stop_reason == Some("tool_use") {
                                tool_fired = true;
                                let mut items: Vec<(usize, ToolUseContent)> = tool_uses.drain().collect();
                                items.sort_by_key(|(idx, _)| *idx);
                                for (_, tu) in items {
                                    on_event(StreamEvent::ToolUse(tu));
                                }
                            }
                        }
                    }
                    "message_stop" => {
                        on_event(StreamEvent::MessageStop {
                            reasoning_content: if thinking_buffer.is_empty() {
                                None
                            } else {
                                Some(thinking_buffer.clone())
                            },
                            usage: last_usage.clone(),
                        });
                    }
                    "error" => {
                        let err = data
                            .get("error")
                            .and_then(|e| e.get("message"))
                            .and_then(Value::as_str)
                            .unwrap_or("Anthropic API error")
                            .to_string();
                        on_event(StreamEvent::Error(err));
                    }
                    _ => {}
                }
                return true;
            }

            if trimmed.is_empty() {
                current_event.clear();
            }
            true
        })
        .await;

        result
    }
}
